### ISSUE TITLE
Addition of a build system generator

### ISSUE THREAD

**Issue by elfring** [namhyung/uftrace#189]
I suggest to reuse a higher level build system than [your current make script](https://github.com/namhyung/uftrace/blob/574f97660f8ea9eec39fcd91d74ea87a6fa9524c/Makefile#L1 "Update candidate") so that powerful checks for software features will become easier.
- [CMake](https://cmake.org/ "Building software with CMake")
- [Autotools](https://www.gnu.org/software/autoconf/autoconf.html#introduction "Building software with Autoconf")

**Comment by namhyung** [comment]
Thanks for the suggestion.  But I'm not familiar with those tools so I worry about the cost of rewriting and maintenance.

**Comment by elfring** [comment]
- Does [a tool like “autoscan”](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.69/html_node/autoscan-Invocation.html "Using autoscan to create configure.ac") show any software dependencies that you find worth for further considerations?
- How do you think about to support more development environments by [the reuse of build system generators](https://cmake.org/cmake/help/v3.10/manual/cmake-generators.7.html#cmake-generators "Listing of CMake generators")?

**Comment by namhyung** [comment]
Yes, those are great features and I definitely need them in the long run.  But maintenace is also important and I don't have much time to study them for now.  Are you willing to volunteer? ;-)

**Comment by gpollo** [comment]
Hi, I've been working with uftrace as part of my research. My code was using `meson` as the build system. Since the current build system is basically raw makefile, I though I'd give a try about converting uftrace to `meson`. The [meson](https://mesonbuild.com/) build system is relatively new, but multiple projects are or has switchted to it like [systemd](https://github.com/systemd/systemd), [pulseaudio](https://github.com/pulseaudio/pulseaudio), [radare2](https://github.com/radareorg/radare2) and [much more](https://mesonbuild.com/Users.html). Having previously worked with `cmake` extensively before, I was very pleased with `meson` clear syntax and how simple it is to use.

Anyway, I have an attempt [here on this branch](https://github.com/gpollo/uftrace/commit/2398086b794c1a6dae967e4f8cf96108ee954f9d). It doesn't support all of the original build options as of yet, but it shouldn't be hard. If there is interest, I wouldn't mind working on it more and sending a PR.

**Comment by honggyukim** [comment]
Hi @gpollo, Thanks very much for bringing this issue again.

I'm actually not familiar with `meson`, but it's good to have such a modern build system because I also think that make is too complicated. I just have a few questions and concerns.
1. What's the benefit of using `cmake`?
2. Does meson have an option that generates `make` if user wants rather than `ninja`?
3. The uftrace is also intended to be used for small embedded systems, but I'm a bit afraid if it's easy to install meson on the device such as raspberry pi and other devices.
4. I found that `meson` has dependencies on Python (version 3.5 or newer) and Ninja (version 1.7 or newer), but that might not be easy to be installed on some small embedded devices.

I think we can maintain both `make` and `meson` build systems at the beginning, then gradually switch it to `meson` when we become more comfortable and time passes that we can expect python 3.5 and ninja 1.7 are widely used even in embedded world.

@namhyung How do you think about it?

**Comment by namhyung** [comment]
In general, I'm happy to see volunteering work on build system improvement.  It seems meson is widely used in various projects so I guess it's already supported by most major embedded system development environments like yocto.  Maybe @chbae have some comments?

One of my concern is that meson seems to require separation of source and build directory.  I think current test code has assumption about the source path being same as build.  If it can run all the tests seamlessly I think we're ok to switch..

**Comment by gpollo** [comment]
@namhyung I'd say the biggest downside of `meson` is that it is quite new and they try to bring new ideas in the build system environment. For example, there is no `make` backend because it is slow and harder to write. For a build system that automatically generates build files, `ninja` is simply a better backend. They also force what they consider good practices. Some things that can be hacked in `cmake` simply cannot be done in `meson` because they don't consider it a good practice.

1. See this section from `meson` [website](https://mesonbuild.com/Comparisons.html). The main benefits of `meson` over `cmake` is simplicity and readability. Having worked on multiples `cmake` projects before, the build definitions can become quite cumbersome. Lots of things seem like "hack" in `cmake` while in `meson` it is well designed. That said, `cmake` is more powerful. If you want a very exotic build system that does lots of things other than building libs and executables, `cmake` can do that.

2. See this section from `meson` [website](https://mesonbuild.com/FAQ.html#why-is-there-not-a-make-backend).

3 & 4. I am not aware of `meson` on embedded systems. Raspberry Pi should be fine though, but Python 3.5 and Ninja, yeah I could see them missing in some case. That said, it is very simple to build Ninja (i.e. no dependencies).

@namhyung I haven't had a look at how tests are executed, but I'll have a look it it later.

**Comment by honggyukim** [comment]
Thanks @gpollo, I'm happy to have a modern build system then.
